runtime/src/main/java/org/corfudb/runtime/object/transactions/AbstractTransactionalContext.java

```diff
@@ -158,37 +152,34 @@ public abstract class AbstractTransactionalContext implements
         AbstractTransactionalContext.log.trace("TXBegin[{}]", this);
     }
 
-    protected <T extends ICorfuSMR<T>> ICorfuSMRSnapshotProxy<T> getAndCacheSnapshotProxy(ICorfuSMRProxyInternal<T> proxy, long ts) {
+    protected <S extends SnapshotGenerator<S> & ConsistencyView> ICorfuSMRSnapshotProxy<S> getAndCacheSnapshotProxy(
+            MVOCorfuCompileProxy<S> proxy, long ts) {
         // TODO: Refactor me to avoid casting on ICorfuSMRProxyInternal type.
-        ICorfuSMRSnapshotProxy<T> snapshotProxy = (ICorfuSMRSnapshotProxy<T>) snapshotProxyMap.get(proxy);
-        final MVOCorfuCompileProxy<T> persistentProxy = (MVOCorfuCompileProxy<T>) proxy;
+        ICorfuSMRSnapshotProxy<S> snapshotProxy = (ICorfuSMRSnapshotProxy<S>) snapshotProxyMap.get(proxy);
         if (snapshotProxy == null) {
-            snapshotProxy = persistentProxy.getUnderlyingMVO().getSnapshotProxy(ts);
+            snapshotProxy = proxy.getUnderlyingMVO().getSnapshotProxy(ts);
             snapshotProxyMap.put(proxy, snapshotProxy);
         }
 
         return snapshotProxy;
     }
 
-    protected void updateKnownStreamPosition(@NonNull ICorfuSMRProxyInternal<?> proxy, long position) {
+    protected void updateKnownStreamPosition(@NonNull MVOCorfuCompileProxy<?> proxy, long position) {
         Long val = knownStreamsPosition.get(proxy.getStreamID());
 
         if (val != null) {
-            if (proxy.isMonotonicStreamAccess()) {
-                Preconditions.checkState(val <= position,
-                        "new stream position %s has decreased from %s", position, val);
-            } else {
-                // This precondition is not valid for monotonic objects since multiple accesses
-                // performed by a transaction may not always see the same stream position.
-                // This can occur if another thread performs accesses at a later snapshot and
-                // interleaves with this transaction.
-                Preconditions.checkState(val == position,
-                        "inconsistent stream positions %s and %s", val, position);
-                return;
-            }
+            // This precondition is not valid for monotonic objects since multiple accesses
+            // performed by a transaction may not always see the same stream position.
+            // This can occur if another thread performs accesses at a later snapshot and
+            // interleaves with this transaction.
+            Preconditions.checkState(val == position,
+                    "inconsistent stream positions %s and %s", val, position);
```

Comment:
Why is a check for objects opened with read committed no longer required here?

Reply:
Fixed.